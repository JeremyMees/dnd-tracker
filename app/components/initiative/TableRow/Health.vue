<script setup lang="ts">
import { reset } from '@formkit/core'
import { INITIATIVE_SHEET } from '~~/constants/provide-keys'
import { useToast } from '~/components/ui/toast/use-toast'
import { hpFunctions, deathSavesFunctions } from '~/utils/dnd-helpers'

const props = defineProps<{ item: InitiativeSheetRow }>()

const { sheet, update } = validateInject(INITIATIVE_SHEET)

const { t } = useI18n()
const { toast } = useToast()

const { hasDeathSaves, checkDeathSaves, addDeathSave, resetDeathSaves } = deathSavesFunctions
const { heal, damage, temp, override, overrideReset } = hpFunctions

const popoverOpen = shallowRef<boolean>(false)
const selectedType = shallowRef<HealthType>('heal')

const hasHp = computed(() => isDefined(props.item.health) && isDefined(props.item.maxHealth))

type HealthType = 'heal' | 'damage' | 'temp' | 'override' | 'override-reset'
interface HealthForm { amount: number }

async function updateRow(row: Partial<InitiativeSheetRow>): Promise<void> {
  if (!sheet.value) return

  const index = getCurrentRowIndex(sheet.value, props.item.id)
  const rows = [...sheet.value.rows]

  if (index === -1 || !rows[index]) return

  rows[index] = { ...rows[index], ...row }

  await update({ rows })
  popoverOpen.value = false
}

async function updateBase(form: HealthForm, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!sheet.value) return

    const { amount } = sanitizeForm<HealthForm>(form)
    const row = {
      ...props.item,
      maxHealth: amount,
      maxHealthOld: undefined,
      health: amount,
    }

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowHealthBase')
    node.setErrors(t('general.error.text'))
  }
}

async function updateOverride(form: HealthForm & { reset?: boolean }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!sheet.value) return

    const { amount, reset } = sanitizeForm<HealthForm & { reset?: boolean }>(form)

    const row = reset || amount === props.item.maxHealthOld
      ? handleHpChanges(props.item.maxHealthOld ?? 0, 'override-reset')
      : handleHpChanges(amount, 'override')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowHealthOverride')
    node.setErrors(t('general.error.text'))
  }
}

async function updateHealth(form: HealthForm, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    const selected = selectedType.value

    if (!sheet.value || !selected) return

    const { amount, type } = sanitizeForm<HealthForm & { type?: HealthType }>({ ...form, type: selected })

    const row = handleHpChanges(amount, type || 'heal')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowHealthUpdate')
    node.setErrors(t('general.error.text'))
  }
}

function handleHpChanges(amount: number, type: HealthType): InitiativeSheetRow {
  const row = { ...props.item }
  const noHp = typeof row.health === 'number' && row.health <= 0

  if (type === 'heal') {
    if (hasDeathSaves(row.type) && noHp) row.deathSaves = resetDeathSaves()

    heal(row, amount)
  }
  else if (type === 'temp') temp(row, amount)
  else if (type === 'override') override(row, amount)
  else if (type === 'override-reset') overrideReset(row, amount)
  else if (type === 'damage') {
    damage(row, amount)

    const downed = typeof row.health === 'number' && row.health <= 0

    // if the health is 0 or less, add 2 death save failures
    if (hasDeathSaves(row.type) && downed) {
      if (!row.deathSaves) {
        row.deathSaves = {
          fail: [false, false, false],
          save: [false, false, false],
        }
      }

      row.deathSaves = addDeathSave(row.deathSaves, 'fail', 2)
    }

    if (row.concentration && !downed) {
      toast({
        title: t('components.initiativeTable.concentration.title'),
        description: t('components.initiativeTable.concentration.text', { name: row.name }),
        variant: 'info',
      })
    }

    if (downed && (row.concentration || row.conditions.length)) {
      row.concentration = false
      row.conditions = []

      toast({
        title: t('components.initiativeTable.downed.title', { name: row.name }),
        description: t('components.initiativeTable.downed.text', { name: row.name }),
        variant: 'info',
      })
    }
  }

  // when user is dies because of going to much in the negative hp
  const dead = (row.health && row.maxHealth && row.health < 0 && Math.abs(row.health) >= row.maxHealth)
  const { failed, saved } = row.deathSaves ? checkDeathSaves(row.deathSaves) : { failed: false, saved: false }

  if (dead || (failed && !saved)) {
    toast({
      title: t('components.initiativeTable.died.title', { name: row.name }),
      description: t('components.initiativeTable.died.textMinHP', { name: row.name }),
      variant: 'info',
    })
  }

  if (!failed && saved) {
    toast({
      title: t('components.initiativeTable.stable.title', { name: row.name }),
      description: t('components.initiativeTable.stable.textDeathSaves', { name: row.name }),
      variant: 'info',
    })
  }

  // when health is an negative number change it to 0
  const resetNegative = sheet.value?.settings?.negative === false
  if (resetNegative && row.health && row.health < 0) row.health = 0

  return row
}
</script>

<template>
  <div>
    <UiPopover v-model:open="popoverOpen">
      <UiPopoverTrigger as-child>
        <button
          data-test-trigger
          :class="{
            'bg-destructive/20 p-2 rounded-lg w-fit': isDefined(item.health) && item.health <= 0,
          }"
          class="flex flex-col gap-y-1"
        >
          <div class="flex items-center gap-x-1">
            <Icon
              v-if="!isDefined(item.health) && item.type !== 'lair'"
              data-test-empty
              name="tabler:plus"
              class="size-5 min-w-5 text-foreground/10"
              aria-hidden="true"
            />
            <span
              v-else
              data-test-health
              :class="{ 'text-destructive': isDefined(item.health) && item.health <= 0 }"
            >
              {{ item.health }}
            </span>
            <span
              v-if="isDefined(item.health) && item.tempHealth"
              v-tippy="$t('general.temp')"
              data-test-temp
              class="text-warning text-xs"
            >
              +{{ item.tempHealth }}
            </span>
          </div>
          <span
            v-if="item.maxHealth !== item.health"
            data-test-max
            class="text-2xs text-muted-foreground whitespace-nowrap"
          >
            {{ $t('general.max') }}: {{ item.maxHealth }}
          </span>
        </button>
      </UiPopoverTrigger>
      <UiPopoverContent>
        <div
          v-if="isDefined(item.health) && isDefined(item.maxHealth)"
          class="flex flex-wrap gap-x-1 gap-y-2 pb-6 items-start justify-center"
        >
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.current') }}
            </p>
            <p
              class="head-2"
              :class="{ 'text-destructive': item.health < 1 }"
            >
              {{ item.health || 0 }}
            </p>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.max') }}
            </p>
            <div class="flex gap-1 items-start justify-center">
              <p
                class="head-2"
                :class="[!item.maxHealthOld ? undefined : item.maxHealthOld < item.maxHealth ? 'text-success' : 'text-destructive']"
              >
                {{ item.maxHealth || 0 }}
              </p>
              <p
                v-if="isDefined(item.maxHealthOld)"
                class="text-sm"
              >
                ({{ item.maxHealthOld }})
              </p>
            </div>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.temp') }}
            </p>
            <p class="head-2">
              {{ item.tempHealth || 0 }}
            </p>
          </div>
        </div>
        <FormKit
          v-if="hasHp"
          id="InitiativeRowHealthUpdate"
          type="form"
          :actions="false"
          @submit="updateHealth"
        >
          <FormDiceRollInput />
          <div class="flex items-center gap-x-2">
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-success flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'heal'"
            >
              <Icon
                name="tabler:heart"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.heal') }}
            </FormKit>
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-warning flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'temp'"
            >
              <Icon
                name="tabler:plus"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.temp') }}
            </FormKit>
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-destructive flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'damage'"
            >
              <Icon
                name="tabler:sword"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.damage') }}
            </FormKit>
          </div>
        </FormKit>
        <UiSeparator
          v-if="hasHp"
          class="my-6 bg-muted-foreground"
        />
        <FormKit
          id="InitiativeRowHealthBase"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateBase"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.baseFieldLabel', { field: 'HP' })"
            :help="$t('components.inputs.baseFieldHelp', { field: 'HP' })"
            validation="required|between:1,1000|number"
          />
        </FormKit>
        <UiSeparator
          v-if="hasHp"
          class="my-6 bg-muted-foreground"
        />
        <FormKit
          v-if="hasHp"
          id="InitiativeRowHealthOverride"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateOverride"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.overrideFieldLabel', { field: 'HP' })"
            :help="$t('components.inputs.optionalFieldHelp', { field: 'HP' })"
            validation="required|between:1,1000|number"
            :suffix-icon="item.maxHealthOld ? 'tabler:player-skip-back' : undefined"
            @suffix-icon-click="(node: FormNode) => updateOverride({ reset: true, amount: item.maxHealthOld || 0 }, node)"
          />
        </FormKit>
      </UiPopoverContent>
    </UiPopover>
  </div>
</template>
