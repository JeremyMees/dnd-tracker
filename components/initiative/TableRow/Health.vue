<script setup lang="ts">
import { reset } from '@formkit/core'
import { useToast } from '~/components/ui/toast/use-toast'
import { hpFunctions, deathSavesFunctions } from '~/utils/dnd-helpers'

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

const { t } = useI18n()
const { toast } = useToast()

const { hasDeathSaves, checkDeathSaves, addDeathSave, resetDeathSaves } = deathSavesFunctions
const { heal, damage, temp, override, overrideReset } = hpFunctions

const popoverOpen = ref<boolean>(false)

type HealthType = 'heal' | 'damage' | 'temp' | 'override' | 'override-reset'
interface HealthForm { amount: number }

async function updateRow(row: Partial<InitiativeSheetRow>): Promise<void> {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1) return

  rows[index] = { ...rows[index], ...row }

  await props.update({ rows })
  popoverOpen.value = false
}

async function updateOverride(form: HealthForm & { reset?: boolean }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet) return

    const { amount, reset } = sanitizeForm<HealthForm & { reset?: boolean }>(form)

    const row = reset || amount === props.item.maxHealthOld
      ? handleHpChanges(props.item.maxHealthOld ?? 0, 'override-reset')
      : handleHpChanges(amount, 'override')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowHealthOverride')
    node.setErrors(t('general.error.text'))
  }
}

async function updateHealth(form: HealthForm & { type?: HealthType }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet || !form.type) return

    const { amount, type } = sanitizeForm<HealthForm & { type?: HealthType }>(form)

    const row = handleHpChanges(amount, type || 'heal')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowHealthUpdate')
    node.setErrors(t('general.error.text'))
  }
}

function handleHpChanges(amount: number, type: HealthType): InitiativeSheetRow {
  const row = { ...props.item }
  const noHp = typeof row.health === 'number' && row.health <= 0

  if (type === 'heal') {
    if (hasDeathSaves(row.type) && noHp) row.deathSaves = resetDeathSaves()

    heal(row, amount)
  }
  else if (type === 'temp') temp(row, amount)
  else if (type === 'override') override(row, amount)
  else if (type === 'override-reset') overrideReset(row, amount)
  else if (type === 'damage') {
    // if the health is 0 or less, add 2 death save failures
    if (hasDeathSaves(row.type) && noHp) row.deathSaves = addDeathSave(row.deathSaves!, 'fail', 2)

    damage(row, amount)

    const downed = typeof row.health === 'number' && row.health <= 0

    if (row.concentration && !downed) {
      toast({
        title: t('components.initiativeTable.concentration.title'),
        description: t('components.initiativeTable.concentration.text', { name: row.name }),
        variant: 'info',
      })
    }

    if (downed && (row.concentration || row.conditions.length)) {
      row.concentration = false
      row.conditions = []

      toast({
        title: t('components.initiativeTable.downed.title', { name: row.name }),
        description: t('components.initiativeTable.downed.text', { name: row.name }),
        variant: 'info',
      })
    }
  }

  // when user is dies because of going to much in the negative hp
  const dead = (row.health && row.maxHealth && row.health < 0 && Math.abs(row.health) >= row.maxHealth)
  const { failed, saved } = row.deathSaves ? checkDeathSaves(row.deathSaves) : { failed: false, saved: false }

  if (dead || (failed && !saved)) {
    toast({
      title: t('components.initiativeTable.died.title', { name: row.name }),
      description: t('components.initiativeTable.died.textMinHP', { name: row.name }),
      variant: 'info',
    })
  }

  if (!failed && saved) {
    toast({
      title: t('components.initiativeTable.stable.title', { name: row.name }),
      description: t('components.initiativeTable.stable.textDeathSaves', { name: row.name }),
      variant: 'info',
    })
  }

  // when health is an negative number change it to 0
  if (row.health && row.health < 0) row.health = 0

  return row
}
</script>

<template>
  <div>
    <UiPopover v-model:open="popoverOpen">
      <UiPopoverTrigger as-child>
        <button
          :class="{
            'bg-destructive/20 p-2 rounded-lg w-fit': isDefined(item.health) && item.health === 0,
          }"
          class="flex flex-col gap-y-1"
        >
          <div class="flex items-center gap-x-1">
            <Icon
              v-if="!isDefined(item.health) && item.type !== 'lair'"
              name="tabler:plus"
              class="size-5 min-w-5 text-foreground/10"
              aria-hidden="true"
            />
            <span
              v-else
              :class="{ 'text-destructive': item.health === 0 }"
            >
              {{ item.health }}
            </span>
            <span
              v-if="isDefined(item.health) && item.tempHealth"
              v-tippy="$t('general.temp')"
              class="text-warning text-xs"
            >
              +{{ item.tempHealth }}
            </span>
          </div>
          <span
            v-if="item.maxHealth !== item.health"
            class="body-extra-small text-muted-foreground whitespace-nowrap"
          >
            {{ $t('general.max') }}: {{ item.maxHealth }}
          </span>
        </button>
      </UiPopoverTrigger>
      <UiPopoverContent>
        <div
          v-if="isDefined(item.health) && isDefined(item.maxHealth)"
          class="flex flex-wrap gap-x-1 gap-y-2 pb-6 items-start justify-center"
        >
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.current') }}
            </p>
            <p
              class="head-2"
              :class="{ 'text-destructive': item.health < 1 }"
            >
              {{ item.health || 0 }}
            </p>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.max') }}
            </p>
            <div class="flex gap-1 items-start justify-center">
              <p
                class="head-2"
                :class="[!item.maxHealthOld ? undefined : item.maxHealthOld < item.maxHealth ? 'text-success' : 'text-destructive']"
              >
                {{ item.maxHealth || 0 }}
              </p>
              <p
                v-if="item.maxHealthOld === 0 || item.maxHealthOld"
                class="body-small"
              >
                ({{ item.maxHealthOld }})
              </p>
            </div>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.temp') }}
            </p>
            <p class="head-2">
              {{ item.tempHealth || 0 }}
            </p>
          </div>
        </div>
        <FormKit
          id="InitiativeRowHealthUpdate"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateHealth"
        >
          <FormDiceRollInput />
          <FormKit
            type="togglebuttons"
            name="type"
            value="damage"
            enforced
            :options="[
              { label: $t('actions.heal'), value: 'heal', icon: 'tabler:heart' },
              { label: $t('actions.temp'), value: 'temp', icon: 'tabler:plus' },
              { label: $t('actions.damage'), value: 'damage', icon: 'tabler:sword' },
            ]"
            validation="required"
            input-class="$remove:p-2 $remove:border-4 $remove:aria-pressed:border-primary $remove:focus-within:border-primary group"
            option-class="$remove:w-10 $remove:h-10"
          >
            <template #default="{ option }">
              <div
                class="flex items-center gap-x-1 rounded-lg p-2 border-2 border-transparent"
                :class="{
                  'bg-destructive/20 group-aria-pressed:bg-destructive/40 group-aria-pressed:border-destructive': option.value === 'damage',
                  'bg-success/20 group-aria-pressed:bg-success/40 group-aria-pressed:border-success': option.value === 'heal',
                  'bg-warning/20 group-aria-pressed:bg-warning/40 group-aria-pressed:border-warning': option.value === 'temp',
                }"
              >
                <Icon
                  :name="option.icon"
                  class="size-4 min-w-4 text-foreground"
                  aria-hidden="true"
                />
                <span class="text-xs text-muted-foreground group-aria-pressed:!text-foreground">
                  {{ option.label }}
                </span>
              </div>
            </template>
          </FormKit>
        </FormKit>
        <UiSeparator class="my-6 bg-muted-foreground" />
        <FormKit
          id="InitiativeRowHealthOverride"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateOverride"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.overrideFieldLabel', { field: 'HP' })"
            :help="$t('components.inputs.optionalFieldHelp', { field: 'HP' })"
            validation="required|between:1,1000|number"
            :suffix-icon="item.maxHealthOld ? 'tabler:player-skip-back' : undefined"
            @suffix-icon-click="(node: FormNode) => updateOverride({ reset: true, amount: item.maxHealthOld || 0 }, node)"
          />
        </FormKit>
      </UiPopoverContent>
    </UiPopover>
  </div>
</template>
