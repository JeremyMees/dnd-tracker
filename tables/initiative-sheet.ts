import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
