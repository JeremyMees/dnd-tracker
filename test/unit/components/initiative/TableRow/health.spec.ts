import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import Health from '~/components/initiative/TableRow/Health.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface HealthTestMethods {
  handleHpChanges: (amount: number, type: 'heal' | 'damage' | 'temp' | 'override' | 'override-reset') => InitiativeSheetRow
  updateRow: (row: Partial<InitiativeSheetRow>) => Promise<void>
}

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()
const mockToast = vi.fn()

vi.mock('~/components/ui/toast/use-toast', () => ({
  useToast: () => ({
    toast: mockToast,
  }),
}))

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row health', async () => {
  beforeEach(() => {
    mockUpdate.mockClear()
    mockToast.mockClear()
  })

  it('Should match snapshot', async () => {
    const component = await mountSuspended(Health, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should display health values correctly', async () => {
    const health = 10
    const maxHealth = 20
    const maxHealthOld = 15
    const tempHealth = 5

    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: { ...props.item, health, maxHealth, maxHealthOld, tempHealth },
      },
    })

    expect(component.get('[data-test-health]').text()).toBe(health.toString())
    expect(component.get('[data-test-max]').text()).toContain(maxHealth.toString())
    expect(component.get('[data-test-temp]').text()).toContain(tempHealth.toString())
  })

  it('Should show destructive styling when health is 0', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: { ...props.item, health: 0 },
      },
    })

    expect(component.get('[data-test-trigger]').classes()).toContain('bg-destructive/20')
    expect(component.get('[data-test-health]').classes()).toContain('text-destructive')
  })

  it('Should show plus icon when health is not defined', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: { ...props.item, health: undefined },
      },
    })

    expect(component.get('[data-test-empty]').isVisible()).toBeTruthy()
  })

  it('Should handle death saves when health reaches 0', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 5,
          type: 'player',
          deathSaves: {
            fail: [false, false, false],
            save: [false, false, false],
          },
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(10, 'damage')
    await vm.updateRow(updatedRow)

    expect(mockUpdate).toHaveBeenCalled()
    const payload = mockUpdate.mock.calls[0]?.[0] as { rows: InitiativeSheetRow[] }
    expect(payload).toBeDefined()
    const resultRow = payload.rows[0]
    expect(resultRow).toBeDefined()
    expect(resultRow?.deathSaves?.fail).toEqual([true, true, false])
  })

  it('Should remove concentration and conditions when health reaches 0', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 5,
          concentration: true,
          conditions: [{ name: 'blinded', desc: 'Cannot see' }],
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(10, 'damage')
    await vm.updateRow(updatedRow)

    expect(mockUpdate).toHaveBeenCalled()
    const payload = mockUpdate.mock.calls[0]?.[0] as { rows: InitiativeSheetRow[] }
    expect(payload).toBeDefined()
    const resultRow = payload.rows[0]
    expect(resultRow).toBeDefined()
    expect(resultRow!.health).toBe(0)
    expect(resultRow!.concentration).toBe(false)
    expect(resultRow!.conditions).toHaveLength(0)
  })

  it('Should show concentration toast when taking damage with concentration', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 20,
          concentration: true,
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(5, 'damage')
    await vm.updateRow(updatedRow)

    expect(mockToast).toHaveBeenCalledWith({
      title: expect.stringMatching('components.initiativeTable.concentration.title'),
      description: expect.stringMatching('components.initiativeTable.concentration.text'),
      variant: 'info',
    })
  })

  it('Should show downed toast when health reaches 0', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 5,
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(10, 'damage')
    await vm.updateRow(updatedRow)

    expect(mockToast).toHaveBeenCalledWith({
      title: expect.stringMatching('components.initiativeTable.downed.title'),
      description: expect.stringMatching('components.initiativeTable.downed.text'),
      variant: 'info',
    })
  })

  it('Should show death toast when health goes below negative max health', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 10,
          maxHealth: 20,
          type: 'player',
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(30, 'damage')
    await vm.updateRow(updatedRow)

    expect(mockToast).toHaveBeenCalledWith({
      title: expect.stringMatching('components.initiativeTable.died.title'),
      description: expect.stringMatching('components.initiativeTable.died.textMinHP'),
      variant: 'info',
    })
  })

  it('Should show stable toast when death saves are successful', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 0,
          maxHealth: 20,
          type: 'player',
          deathSaves: {
            save: [true, true, true],
            fail: [false, false, false],
          },
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(10, 'temp')
    await vm.updateRow(updatedRow)

    expect(component.get('[data-test-health]').text()).toBe('0')
  })

  it('Should set health to 0 when negative values are not allowed', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 10,
          maxHealth: 20,
          tempHealth: 0,
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(15, 'damage')
    await vm.updateRow(updatedRow)

    expect(updatedRow.health).toBe(0)
  })

  it('Should allow negative health when negative values are allowed', async () => {
    const component = await mountSuspended(Health, {
      props: {
        ...props,
        item: {
          ...props.item,
          health: 10,
          maxHealth: 20,
          tempHealth: 0,
        },
        sheet: {
          ...sheet,
          settings: {
            ...sheet.settings,
            negative: true,
          },
        },
      },
    })

    const vm = component.vm as unknown as HealthTestMethods
    const updatedRow = vm.handleHpChanges(15, 'damage')
    await vm.updateRow(updatedRow)

    expect(updatedRow.health).toBe(-5)
  })
})
