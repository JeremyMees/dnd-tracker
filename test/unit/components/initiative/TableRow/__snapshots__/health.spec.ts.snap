// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row health > Should match snapshot 1`] = `
"<div><button id="reka-popover-trigger-v-0-0-0" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed" data-test-trigger="" class="flex flex-col gap-y-1">
    <div class="flex items-center gap-x-1"><span data-test-health="" class="">10</span><span data-test-temp="" class="text-warning text-xs"> +5</span></div><span data-test-max="" class="text-2xs text-muted-foreground whitespace-nowrap">general.max: 100</span>
  </button>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
